2016/submission/ref-util.pod

```diff
@@ -18,49 +30,101 @@ maintain: Checking the validity of references using the C<ref> function.
+replaced by something like this:
 
-    # before
+    # In a module, and tested
+    sub is_plain_hash {
+        return ref $_[0] eq 'HASH';
+    }
+
+   is_plain_hash($plate_contents)
+       or croak('Plate contents not passed as a hashref');
+
+Since under strict a typo in C<is_plain_hash> would cause a compile time error
+any fat-fingered elf problems would be caught. However, that's just introduced
+another subroutine call, and since these ref checks are used all over Santa's
+code base, the cost of doing that adds up!
+
+The second problem with the pattern was the fact that it was strictly doing more
+work that it needed to.  Consider what the following was actually doing:
+
+    ref $args eq 'HASH'
+
+Under the hood perl is inspecting bits on the reference to see if it's a C<HV>
+(a C<hash value>), then building a standard SV string containing the four
+characters C<HASH>, then in a separate operation comparing that string character
+by character with the C<HASH> string was passed in in the code.  That, of
+course, all happens really quickly, but it's more work (and more Perl virtual
+ops) than need happen just to tell if this was a hashref or not!
+
+So the code review had found two problems:  One, an abstraction was needed to
+make it more reliable.  Two, the code needed to involve less stuff to make it
+faster.
+
+=head2 The Voice of an Angel
+
+Thank goodness the Wise Old Elf read a L<CPAN Weekly|http://cpan-weekly.org/>
+email about this very pattern! Because of this he decided to try out a new
+module, L<Ref::Util>.
+
+L<Ref::Util> provides a set of helpful functions to determine what kind of
+reference a variable is. It abstracts several awkward ref checking patterns into
+new keywords.  So, as our above example:
 
     ref $args eq 'HASH'
         or croak('Sled expects arguments as a hashref');
 
-    # after
+Can be written as:
+
     use Ref::Util qw<is_plain_hash>;
 
     is_plain_hash($args)
         or croak('Sled expects arguments as a hashref');
 
 Or maybe we want to check for a blessed array reference, making sure it
-isn't accidentally blessed:
+isn't accidentally blessed? The following:
 
-    use Carp qw<croak>;
-
-    # before
     use Scalar::Util qw<blessed reftype>;
 
     blessed $args && ref $args && reftype($args) eq 'ARRAY'
         or Carp::croak('Uh oh, we require a blessed reference');
 
-    # after
+Can much more succinctly be written with the new C<is_blessed_arrayref> keyword:
+
     use Ref::Util qw<is_blessed_arrayref>;
 
     is_blessed_arrayref($args)
         or Carp::croak('Uh oh, we require a blessed reference');
 
 Not writing a lot of code but still being accurate is something Santa
-appreciates, but that's not the best part.
+appreciates, but that's not the best part....
+
+The Ref::Util module introduces new B<keywords> not just new I<functions> using
+Perl's pluggable keyword infrastructure.  The guts of the checks are implemented
```

Comment:
That's not correct. It actually does it in a simpler (and at the same time, more complicated) way: It introduced, during compile time, new ops (as custom op) in the op tree for Perl, which are then implemented with low-level C that interacts with the Perl internals directly, without any XS macros or XS layer.

When you call `ref $foo eq 'HASH'` there are several ops that come into play:
```
...
6     <2> seq vK/2 ->7
4        <1> ref[t2] sK/1 ->5
3              <#> gvsv[*foo] s ->4
5        <$> const[PV "HASH"] s ->6
```

Each of these is an operation that perl has to do (create a list, get the value, call `ref`, etc.). However, when you use Ref::Util, when `is_hashref` gets encountered, it is then replaced with the following op:

```
4     <1> is_hashref vK/1 ->5
3           <#> gvsv[*foo] s ->4
```

This is much faster. Additionally, this single op is not XS, but low-level C as explained above.

Further more Ref::Util only uses actual XS implementation when you get around the compilation phase by calling a function dynamically:

```
use Ref::Util;
Ref::Util::->can("is_hashref")->($foo);
```

In this case, the XS code is triggered instead (because it can't be replaced), but that is still faster than the `ref` call.

(This seemed very technical so I didn't get into that.)